Ładniejszy komunikat dla błędu połączenia przez nieważny cert, gdy w telefonie ustawiony jest zły czas

> Caused by: java.security.cert.CertPathValidatorException: Response is unreliable: its validity interval is out-of-date

pełny przykładowy stack:
```
javax.net.ssl.SSLHandshakeException: Chain validation failed
at com.android.org.conscrypt.SSLUtils.toSSLHandshakeException(SSLUtils.java:362)
at com.android.org.conscrypt.ConscryptEngine.convertException(ConscryptEngine.java:1134)
at com.android.org.conscrypt.ConscryptEngine.readPlaintextData(ConscryptEngine.java:1089)
at com.android.org.conscrypt.ConscryptEngine.unwrap(ConscryptEngine.java:876)
at com.android.org.conscrypt.ConscryptEngine.unwrap(ConscryptEngine.java:747)
at com.android.org.conscrypt.ConscryptEngine.unwrap(ConscryptEngine.java:712)
at com.android.org.conscrypt.ConscryptEngineSocket$SSLInputStream.processDataFromSocket(ConscryptEngineSocket.java:858)
at com.android.org.conscrypt.ConscryptEngineSocket$SSLInputStream.access$100(ConscryptEngineSocket.java:731)
at com.android.org.conscrypt.ConscryptEngineSocket.doHandshake(ConscryptEngineSocket.java:241)
at com.android.org.conscrypt.ConscryptEngineSocket.startHandshake(ConscryptEngineSocket.java:220)
at okhttp3.internal.connection.RealConnection.connectTls(RealConnection.kt:379)
at okhttp3.internal.connection.RealConnection.establishProtocol(RealConnection.kt:337)
at okhttp3.internal.connection.RealConnection.connect(RealConnection.kt:209)
at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.kt:226)
at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.kt:106)
at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.kt:74)
at okhttp3.internal.connection.RealCall.initExchange$okhttp(RealCall.kt:255)
at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:32)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
at io.github.wulkanowy.sdk.scrapper.interceptor.StudentCookieInterceptor.intercept(StudentCookieInterceptor.kt:32)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
at io.github.wulkanowy.sdk.scrapper.interceptor.HttpErrorInterceptor.intercept(HttpErrorInterceptor.kt:13)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
at io.github.wulkanowy.sdk.scrapper.interceptor.UserAgentInterceptor.intercept(UserAgentInterceptor.kt:18)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
at io.github.wulkanowy.sdk.scrapper.interceptor.AutoLoginInterceptor.intercept(AutoLoginInterceptor.kt:55)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
at io.github.wulkanowy.sdk.scrapper.interceptor.ErrorInterceptor.intercept(ErrorInterceptor.kt:25)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
at okhttp3.logging.HttpLoggingInterceptor.intercept(HttpLoggingInterceptor.kt:221)
at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:517)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
at java.lang.Thread.run(Thread.java:923)
Caused by: java.security.cert.CertificateException: Chain validation failed
at com.android.org.conscrypt.TrustManagerImpl.verifyChain(TrustManagerImpl.java:724)
at com.android.org.conscrypt.TrustManagerImpl.checkTrustedRecursive(TrustManagerImpl.java:554)
at com.android.org.conscrypt.TrustManagerImpl.checkTrustedRecursive(TrustManagerImpl.java:575)
at com.android.org.conscrypt.TrustManagerImpl.checkTrustedRecursive(TrustManagerImpl.java:620)
at com.android.org.conscrypt.TrustManagerImpl.checkTrusted(TrustManagerImpl.java:510)
at com.android.org.conscrypt.TrustManagerImpl.checkTrusted(TrustManagerImpl.java:428)
at com.android.org.conscrypt.TrustManagerImpl.getTrustedChainForServer(TrustManagerImpl.java:356)
at android.security.net.config.NetworkSecurityTrustManager.checkServerTrusted(NetworkSecurityTrustManager.java:94)
at android.security.net.config.RootTrustManager.checkServerTrusted(RootTrustManager.java:90)
at com.android.org.conscrypt.ConscryptEngineSocket$2.checkServerTrusted(ConscryptEngineSocket.java:163)
at com.android.org.conscrypt.Platform.checkServerTrusted(Platform.java:254)
at com.android.org.conscrypt.ConscryptEngine.verifyCertificateChain(ConscryptEngine.java:1644)
at com.android.org.conscrypt.NativeCrypto.ENGINE_SSL_read_direct(Native Method)
at com.android.org.conscrypt.NativeSsl.readDirectByteBuffer(NativeSsl.java:568)
at com.android.org.conscrypt.ConscryptEngine.readPlaintextDataDirect(ConscryptEngine.java:1095)
at com.android.org.conscrypt.ConscryptEngine.readPlaintextData(ConscryptEngine.java:1079)
... 39 more
Caused by: java.security.cert.CertPathValidatorException: Response is unreliable: its validity interval is out-of-date
at sun.security.provider.certpath.PKIXMasterCertPathValidator.validate(PKIXMasterCertPathValidator.java:135)
at sun.security.provider.certpath.PKIXCertPathValidator.validate(PKIXCertPathValidator.java:222)
at sun.security.provider.certpath.PKIXCertPathValidator.validate(PKIXCertPathValidator.java:140)
at sun.security.provider.certpath.PKIXCertPathValidator.engineValidate(PKIXCertPathValidator.java:79)
at java.security.cert.CertPathValidator.validate(CertPathValidator.java:301)
at com.android.org.conscrypt.TrustManagerImpl.verifyChain(TrustManagerImpl.java:720)
... 54 more
Caused by: java.security.cert.CertPathValidatorException: Response is unreliable: its validity interval is out-of-date
at sun.security.provider.certpath.OCSPResponse.verify(OCSPResponse.java:619)
at sun.security.provider.certpath.RevocationChecker.checkOCSP(RevocationChecker.java:709)
at sun.security.provider.certpath.RevocationChecker.check(RevocationChecker.java:363)
at sun.security.provider.certpath.RevocationChecker.check(RevocationChecker.java:337)
at sun.security.provider.certpath.PKIXMasterCertPathValidator.validate(PKIXMasterCertPathValidator.java:125)
... 59 more
Suppressed: java.security.cert.CertPathValidatorException: Could not determine revocation status
at sun.security.provider.certpath.RevocationChecker.buildToNewKey(RevocationChecker.java:1092)
at sun.security.provider.certpath.RevocationChecker.verifyWithSeparateSigningKey(RevocationChecker.java:910)
at sun.security.provider.certpath.RevocationChecker.checkCRLs(RevocationChecker.java:577)
at sun.security.provider.certpath.RevocationChecker.checkCRLs(RevocationChecker.java:465)
at sun.security.provider.certpath.RevocationChecker.check(RevocationChecker.java:394)
... 61 more
```

inny stack:
```
javax.net.ssl.SSLHandshakeException: Unacceptable certificate: CN=Certum Trusted Network CA, OU=Certum Certification Authority, O=Unizeto Technologies S.A., C=PL
	at com.android.org.conscrypt.SSLUtils.toSSLHandshakeException(SSLUtils.java:363)
	at com.android.org.conscrypt.ConscryptEngine.convertException(ConscryptEngine.java:1134)
	at com.android.org.conscrypt.ConscryptEngine.readPlaintextData(ConscryptEngine.java:1089)
	at com.android.org.conscrypt.ConscryptEngine.unwrap(ConscryptEngine.java:876)
	at com.android.org.conscrypt.ConscryptEngine.unwrap(ConscryptEngine.java:747)
	at com.android.org.conscrypt.ConscryptEngine.unwrap(ConscryptEngine.java:712)
	at com.android.org.conscrypt.ConscryptEngineSocket$SSLInputStream.processDataFromSocket(ConscryptEngineSocket.java:858)
	at com.android.org.conscrypt.ConscryptEngineSocket$SSLInputStream.access$100(ConscryptEngineSocket.java:731)
	at com.android.org.conscrypt.ConscryptEngineSocket.doHandshake(ConscryptEngineSocket.java:241)
	at com.android.org.conscrypt.ConscryptEngineSocket.startHandshake(ConscryptEngineSocket.java:220)
	at okhttp3.internal.connection.RealConnection.connectTls(RealConnection.kt:379)
	at okhttp3.internal.connection.RealConnection.establishProtocol(RealConnection.kt:337)
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.kt:209)
	at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.kt:226)
	at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.kt:106)
	at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.kt:74)
	at okhttp3.internal.connection.RealCall.initExchange$okhttp(RealCall.kt:255)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:32)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at io.github.wulkanowy.sdk.scrapper.interceptor.HttpErrorInterceptor.intercept(HttpErrorInterceptor.kt:13)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at io.github.wulkanowy.sdk.scrapper.interceptor.UserAgentInterceptor.intercept(UserAgentInterceptor.kt:18)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at io.github.wulkanowy.sdk.scrapper.interceptor.AutoLoginInterceptor.intercept(AutoLoginInterceptor.kt:55)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at io.github.wulkanowy.sdk.scrapper.interceptor.ErrorInterceptor.intercept(ErrorInterceptor.kt:25)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.logging.HttpLoggingInterceptor.intercept(HttpLoggingInterceptor.kt:221)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
	at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:517)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:920)
Caused by: java.security.cert.CertificateException: Unacceptable certificate: CN=Certum Trusted Network CA, OU=Certum Certification Authority, O=Unizeto Technologies S.A., C=PL
	at com.android.org.conscrypt.TrustManagerImpl.checkTrustedRecursive(TrustManagerImpl.java:609)
	at com.android.org.conscrypt.TrustManagerImpl.checkTrustedRecursive(TrustManagerImpl.java:638)
	at com.android.org.conscrypt.TrustManagerImpl.checkTrusted(TrustManagerImpl.java:505)
	at com.android.org.conscrypt.TrustManagerImpl.checkTrusted(TrustManagerImpl.java:425)
	at com.android.org.conscrypt.TrustManagerImpl.getTrustedChainForServer(TrustManagerImpl.java:353)
	at android.security.net.config.NetworkSecurityTrustManager.checkServerTrusted(NetworkSecurityTrustManager.java:94)
	at android.security.net.config.RootTrustManager.checkServerTrusted(RootTrustManager.java:90)
	at com.android.org.conscrypt.ConscryptEngineSocket$2.checkServerTrusted(ConscryptEngineSocket.java:163)
	at com.android.org.conscrypt.Platform.checkServerTrusted(Platform.java:255)
	at com.android.org.conscrypt.ConscryptEngine.verifyCertificateChain(ConscryptEngine.java:1638)
	at com.android.org.conscrypt.NativeCrypto.ENGINE_SSL_read_direct(Native Method)
	at com.android.org.conscrypt.NativeSsl.readDirectByteBuffer(NativeSsl.java:569)
	at com.android.org.conscrypt.ConscryptEngine.readPlaintextDataDirect(ConscryptEngine.java:1095)
	at com.android.org.conscrypt.ConscryptEngine.readPlaintextData(ConscryptEngine.java:1079)
	... 37 more
Caused by: java.security.cert.CertificateNotYetValidException: Certificate not valid until Wed Oct 22 14:07:37 GMT+02:00 2008 (compared to Mon Dec 31 14:58:47 GMT+01:00 2007)
	at com.android.org.conscrypt.OpenSSLX509Certificate.checkValidity(OpenSSLX509Certificate.java:264)
	at com.android.org.conscrypt.OpenSSLX509Certificate.checkValidity(OpenSSLX509Certificate.java:255)
	at com.android.org.conscrypt.TrustManagerImpl.checkTrustedRecursive(TrustManagerImpl.java:605)
	... 50 more
```

--- a/app/src/main/java/io/github/wulkanowy/ui/base/ErrorDialog.kt
+++ b/app/src/main/java/io/github/wulkanowy/ui/base/ErrorDialog.kt
@@ -16,15 +16,7 @@ import com.google.android.material.dialog.MaterialAlertDialogBuilder
 import dagger.hilt.android.AndroidEntryPoint
 import io.github.wulkanowy.R
 import io.github.wulkanowy.databinding.DialogErrorBinding
-import io.github.wulkanowy.sdk.exception.FeatureNotAvailableException
-import io.github.wulkanowy.sdk.scrapper.exception.FeatureDisabledException
-import io.github.wulkanowy.sdk.scrapper.exception.ServiceUnavailableException
 import io.github.wulkanowy.utils.*
-import okhttp3.internal.http2.StreamResetException
-import java.io.InterruptedIOException
-import java.net.ConnectException
-import java.net.SocketTimeoutException
-import java.net.UnknownHostException
 import javax.inject.Inject
 
 @AndroidEntryPoint
@@ -79,18 +71,6 @@ class ErrorDialog : DialogFragment() {
         setOnShowListener {
             getButton(AlertDialog.BUTTON_NEUTRAL).isEnabled = isErrorShouldBeReported(error)
         }
-    }
-
-    private fun isErrorShouldBeReported(error: Throwable): Boolean = when (error) {
-        is UnknownHostException,
-        is InterruptedIOException,
-        is ConnectException,
-        is StreamResetException,
-        is SocketTimeoutException,
-        is ServiceUnavailableException,
-        is FeatureDisabledException,
-        is FeatureNotAvailableException -> false
-        else -> true
     }
 
     private fun copyErrorToClipboard(errorStacktrace: String) {

--- a/app/src/main/java/io/github/wulkanowy/ui/base/ErrorDialog.kt
+++ b/app/src/main/java/io/github/wulkanowy/ui/base/ErrorDialog.kt
@@ -59,7 +59,7 @@ class ErrorDialog : DialogFragment() {
 
     private fun DialogErrorBinding.bindErrorDetails(error: Throwable) {
         return with(this) {
-            errorDialogHumanizedMessage.text = resources.getString(error)
+            errorDialogHumanizedMessage.text = resources.getErrorString(error)
             errorDialogErrorMessage.text = error.localizedMessage
             errorDialogErrorMessage.isGone = error.localizedMessage.isNullOrBlank()
             errorDialogContent.text = error.stackTraceToString()
@@ -69,7 +69,7 @@ class ErrorDialog : DialogFragment() {
 
     private fun AlertDialog.enableReportButtonIfErrorIsReportable(error: Throwable) {
         setOnShowListener {
-            getButton(AlertDialog.BUTTON_NEUTRAL).isEnabled = isErrorShouldBeReported(error)
+            getButton(AlertDialog.BUTTON_NEUTRAL).isEnabled = error.isShouldBeReported()
         }
     }
 

--- a/app/src/main/java/io/github/wulkanowy/ui/base/ErrorHandler.kt
+++ b/app/src/main/java/io/github/wulkanowy/ui/base/ErrorHandler.kt
@@ -5,7 +5,7 @@ import dagger.hilt.android.qualifiers.ApplicationContext
 import io.github.wulkanowy.data.exceptions.NoCurrentStudentException
 import io.github.wulkanowy.sdk.scrapper.login.BadCredentialsException
 import io.github.wulkanowy.sdk.scrapper.login.PasswordChangeRequiredException
-import io.github.wulkanowy.utils.getString
+import io.github.wulkanowy.utils.getErrorString
 import io.github.wulkanowy.utils.security.ScramblerException
 import timber.log.Timber
 import javax.inject.Inject
@@ -26,7 +26,7 @@ open class ErrorHandler @Inject constructor(@ApplicationContext protected val co
     }
 
     protected open fun proceed(error: Throwable) {
-        showErrorMessage(context.resources.getString(error), error)
+        showErrorMessage(context.resources.getErrorString(error), error)
         when (error) {
             is PasswordChangeRequiredException -> onPasswordChangeRequired(error.redirectUrl)
             is ScramblerException, is BadCredentialsException -> onSessionExpired()

--- a/app/src/main/java/io/github/wulkanowy/ui/modules/dashboard/DashboardFragment.kt
+++ b/app/src/main/java/io/github/wulkanowy/ui/modules/dashboard/DashboardFragment.kt
@@ -176,7 +176,7 @@ class DashboardFragment : BaseFragment<FragmentDashboardBinding>(R.layout.fragme
     }
 
     override fun setErrorDetails(error: Throwable) {
-        binding.dashboardErrorMessage.text = requireContext().resources.getString(error)
+        binding.dashboardErrorMessage.text = requireContext().resources.getErrorString(error)
     }
 
     override fun resetView() {

--- a/app/src/main/java/io/github/wulkanowy/ui/modules/dashboard/DashboardFragment.kt
+++ b/app/src/main/java/io/github/wulkanowy/ui/modules/dashboard/DashboardFragment.kt
@@ -28,10 +28,7 @@ import io.github.wulkanowy.ui.modules.message.MessageFragment
 import io.github.wulkanowy.ui.modules.notificationscenter.NotificationsCenterFragment
 import io.github.wulkanowy.ui.modules.schoolannouncement.SchoolAnnouncementFragment
 import io.github.wulkanowy.ui.modules.timetable.TimetableFragment
-import io.github.wulkanowy.utils.capitalise
-import io.github.wulkanowy.utils.getThemeAttrColor
-import io.github.wulkanowy.utils.openInternetBrowser
-import io.github.wulkanowy.utils.toFormattedString
+import io.github.wulkanowy.utils.*
 import java.time.LocalDate
 import javax.inject.Inject
 
@@ -178,8 +175,8 @@ class DashboardFragment : BaseFragment<FragmentDashboardBinding>(R.layout.fragme
         binding.dashboardErrorContainer.isVisible = show
     }
 
-    override fun setErrorDetails(message: String) {
-        binding.dashboardErrorMessage.text = message
+    override fun setErrorDetails(error: Throwable) {
+        binding.dashboardErrorMessage.text = requireContext().resources.getString(error)
     }
 
     override fun resetView() {

--- a/app/src/main/java/io/github/wulkanowy/ui/modules/dashboard/DashboardPresenter.kt
+++ b/app/src/main/java/io/github/wulkanowy/ui/modules/dashboard/DashboardPresenter.kt
@@ -714,6 +714,7 @@ class DashboardPresenter @Inject constructor(
                 if ((forceRefresh && wasGeneralError) || !forceRefresh) {
                     showContent(false)
                     showErrorView(true)
+                    setErrorDetails(lastError)
                 }
             }
         }

--- a/app/src/main/java/io/github/wulkanowy/ui/modules/dashboard/DashboardView.kt
+++ b/app/src/main/java/io/github/wulkanowy/ui/modules/dashboard/DashboardView.kt
@@ -18,7 +18,7 @@ interface DashboardView : BaseView {
 
     fun showErrorView(show: Boolean)
 
-    fun setErrorDetails(message: String)
+    fun setErrorDetails(error: Throwable)
 
     fun resetView()
 

--- a/app/src/main/java/io/github/wulkanowy/utils/ExceptionExtension.kt
+++ b/app/src/main/java/io/github/wulkanowy/utils/ExceptionExtension.kt
@@ -40,6 +40,23 @@ fun Resources.getString(error: Throwable): String = when (error) {
     else -> R.string.error_unknown
 }.let { getString(it) }
 
+fun isErrorShouldBeReported(error: Throwable): Boolean = when (error) {
+    is UnknownHostException,
+    is SocketException,
+    is SocketTimeoutException,
+    is InterruptedIOException,
+    is ConnectException,
+    is StreamResetException,
+    is ServiceUnavailableException,
+    is FeatureDisabledException,
+    is FeatureNotAvailableException -> false
+    is SSLHandshakeException -> when {
+        isCausedByCertificateNotValidNow(error) -> false
+        else -> true
+    }
+    else -> true
+}
+
 private fun isCausedByCertificateNotValidNow(e: Throwable?): Boolean {
     var exception = e
     do {

--- a/app/src/main/java/io/github/wulkanowy/utils/ExceptionExtension.kt
+++ b/app/src/main/java/io/github/wulkanowy/utils/ExceptionExtension.kt
@@ -19,7 +19,7 @@ import java.security.cert.CertificateExpiredException
 import java.security.cert.CertificateNotYetValidException
 import javax.net.ssl.SSLHandshakeException
 
-fun Resources.getString(error: Throwable): String = when (error) {
+fun Resources.getErrorString(error: Throwable): String = when (error) {
     is UnknownHostException -> R.string.error_no_internet
     is SocketException,
     is SocketTimeoutException,
@@ -34,13 +34,13 @@ fun Resources.getString(error: Throwable): String = when (error) {
     is VulcanException -> R.string.error_unknown_uonet
     is ScrapperException -> R.string.error_unknown_app
     is SSLHandshakeException -> when {
-        isCausedByCertificateNotValidNow(error) -> R.string.error_invalid_device_datetime
+        error.isCausedByCertificateNotValidNow() -> R.string.error_invalid_device_datetime
         else -> R.string.error_timeout
     }
     else -> R.string.error_unknown
 }.let { getString(it) }
 
-fun isErrorShouldBeReported(error: Throwable): Boolean = when (error) {
+fun Throwable.isShouldBeReported(): Boolean = when (this) {
     is UnknownHostException,
     is SocketException,
     is SocketTimeoutException,
@@ -51,24 +51,24 @@ fun isErrorShouldBeReported(error: Throwable): Boolean = when (error) {
     is FeatureDisabledException,
     is FeatureNotAvailableException -> false
     is SSLHandshakeException -> when {
-        isCausedByCertificateNotValidNow(error) -> false
+        isCausedByCertificateNotValidNow() -> false
         else -> true
     }
     else -> true
 }
 
-private fun isCausedByCertificateNotValidNow(e: Throwable?): Boolean {
-    var exception = e
+private fun Throwable?.isCausedByCertificateNotValidNow(): Boolean {
+    var exception = this
     do {
-        if (isCertificateNotValidNow(exception)) return true
+        if (exception.isCertificateNotValidNow()) return true
 
         exception = exception?.cause
     } while (exception != null)
     return false
 }
 
-private fun isCertificateNotValidNow(throwable: Throwable?): Boolean {
-    val isNotYetValid = throwable is CertificateNotYetValidException
-    val isExpired = throwable is CertificateExpiredException
+private fun Throwable?.isCertificateNotValidNow(): Boolean {
+    val isNotYetValid = this is CertificateNotYetValidException
+    val isExpired = this is CertificateExpiredException
     return isNotYetValid || isExpired
 }

--- a/app/src/main/java/io/github/wulkanowy/utils/ResourcesExtension.kt
+++ b/app/src/main/java/io/github/wulkanowy/utils/ResourcesExtension.kt
@@ -15,8 +15,6 @@ import java.net.ConnectException
 import java.net.SocketException
 import java.net.SocketTimeoutException
 import java.net.UnknownHostException
-import java.security.cert.CertPathValidatorException
-import java.security.cert.CertificateException
 import java.security.cert.CertificateExpiredException
 import java.security.cert.CertificateNotYetValidException
 import javax.net.ssl.SSLHandshakeException
@@ -35,15 +33,25 @@ fun Resources.getString(error: Throwable): String = when (error) {
     is FeatureNotAvailableException -> R.string.error_feature_not_available
     is VulcanException -> R.string.error_unknown_uonet
     is ScrapperException -> R.string.error_unknown_app
-    is SSLHandshakeException -> {
-        val isCertException = error.cause is CertificateException
-        val isPatchValidatorException = error.cause?.cause is CertPathValidatorException
-        val isNotYetValid = error.cause?.cause?.cause is CertificateNotYetValidException
-        val isExpired = error.cause?.cause?.cause is CertificateExpiredException
-
-        if (isCertException && isPatchValidatorException && (isNotYetValid || isExpired)) {
-            R.string.error_invalid_device_datetime
-        } else R.string.error_timeout
+    is SSLHandshakeException -> when {
+        isCausedByCertificateNotValidNow(error) -> R.string.error_invalid_device_datetime
+        else -> R.string.error_timeout
     }
     else -> R.string.error_unknown
 }.let { getString(it) }
+
+private fun isCausedByCertificateNotValidNow(e: Throwable?): Boolean {
+    var exception = e
+    do {
+        if (isCertificateNotValidNow(exception)) return true
+
+        exception = exception?.cause
+    } while (exception != null)
+    return false
+}
+
+private fun isCertificateNotValidNow(throwable: Throwable?): Boolean {
+    val isNotYetValid = throwable is CertificateNotYetValidException
+    val isExpired = throwable is CertificateExpiredException
+    return isNotYetValid || isExpired
+}

--- a/app/src/main/java/io/github/wulkanowy/utils/ResourcesExtension.kt
+++ b/app/src/main/java/io/github/wulkanowy/utils/ResourcesExtension.kt
@@ -15,20 +15,35 @@ import java.net.ConnectException
 import java.net.SocketException
 import java.net.SocketTimeoutException
 import java.net.UnknownHostException
+import java.security.cert.CertPathValidatorException
+import java.security.cert.CertificateException
+import java.security.cert.CertificateExpiredException
+import java.security.cert.CertificateNotYetValidException
+import javax.net.ssl.SSLHandshakeException
 
-fun Resources.getString(error: Throwable) = when (error) {
-    is UnknownHostException -> getString(R.string.error_no_internet)
+fun Resources.getString(error: Throwable): String = when (error) {
+    is UnknownHostException -> R.string.error_no_internet
     is SocketException,
     is SocketTimeoutException,
     is InterruptedIOException,
     is ConnectException,
-    is StreamResetException -> getString(R.string.error_timeout)
-    is NotLoggedInException -> getString(R.string.error_login_failed)
-    is PasswordChangeRequiredException -> getString(R.string.error_password_change_required)
-    is ServiceUnavailableException -> getString(R.string.error_service_unavailable)
-    is FeatureDisabledException -> getString(R.string.error_feature_disabled)
-    is FeatureNotAvailableException -> getString(R.string.error_feature_not_available)
-    is VulcanException -> getString(R.string.error_unknown_uonet)
-    is ScrapperException -> getString(R.string.error_unknown_app)
-    else -> getString(R.string.error_unknown)
-}
+    is StreamResetException -> R.string.error_timeout
+    is NotLoggedInException -> R.string.error_login_failed
+    is PasswordChangeRequiredException -> R.string.error_password_change_required
+    is ServiceUnavailableException -> R.string.error_service_unavailable
+    is FeatureDisabledException -> R.string.error_feature_disabled
+    is FeatureNotAvailableException -> R.string.error_feature_not_available
+    is VulcanException -> R.string.error_unknown_uonet
+    is ScrapperException -> R.string.error_unknown_app
+    is SSLHandshakeException -> {
+        val isCertException = error.cause is CertificateException
+        val isPatchValidatorException = error.cause?.cause is CertPathValidatorException
+        val isNotYetValid = error.cause?.cause?.cause is CertificateNotYetValidException
+        val isExpired = error.cause?.cause?.cause is CertificateExpiredException
+
+        if (isCertException && isPatchValidatorException && (isNotYetValid || isExpired)) {
+            R.string.error_invalid_device_datetime
+        } else R.string.error_timeout
+    }
+    else -> R.string.error_unknown
+}.let { getString(it) }

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -767,6 +767,7 @@
 
     <!--Errors-->
     <string name="error_no_internet">No internet connection</string>
+    <string name="error_invalid_device_datetime">An error occurred. Check your device clock</string>
     <string name="error_timeout">Connection to register failed. Servers can be overloaded. Please try again later</string>
     <string name="error_login_failed">Loading data failed. Please try again later</string>
     <string name="error_password_change_required">Register password change required</string>

